enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,885 @@
+equivalent of `circtl pull ...`. It doesn't improve performance or reduces the
+required size, but it means that there is one less component needed (no need
+for skopeo) and reduces the risks: CRI-O will be writing the images to the disk
+itself, so no risk of format mismatch.
+
+We aren't ruling out any of the above possibilities, consider them as
+implementation details, but we think that the last one is the better overall.
+
+#### CVO will need to orchestrate the upgrade activities
+
+To initiate the upgrade the administrator of the cluster changes the
+`ClusterVersion` object is changed like this:
+
+```yaml
+kind: ClusterVersion
+metadata:
+  name: version
+spec:
+  desiredUpdate:
+    version: 4.14.5
+```
+
+The cluster operator will first ask the machine config operator to pin and
+pre-load the release image, creating a `ContainerRuntimeConfig` object similar
+to this:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: ContainerRuntimeConfig
+metadata:
+  name: pin-upgrade
+spec:
+  containerRuntimeConfig:
+    pinnedImages:
+    - quay.io/openshift-release-dev/ocp-release@sha256:...
+```
+
+The machine config operator will react to that pinning and pre-loading the
+image in all the nodes of the cluster, as described in the previous section.
+Once the image is pinned and pre-loaded the cluster version operator will
+inspect it and find out the references to the payload images. It will then also
+pin and pre-load those images, updating the `ContainerRuntimeConfig` object:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: ContainerRuntimeConfig
+metadata:
+  name: pin-upgrade
+spec:
+  containerRuntimeConfig:
+    pinnedImages:
+    - quay.io/openshift-release-dev/ocp-release@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    ...
+```
+
+Once those images are pinned and pre-loaded the upgrade will proceed as usual.
+
+Note that this process will by default work using whatever registry servers are
+configured in the cluster. When it isn't possible to use a registry server, the
+administrator of the cluster will explicitly configure the cluster version
+operator to use an upgrade bundle instead, setting the new
+`spec.desiredUpdate.imageSource.sourceType` field to `Bundle` (the default will
+be `Registry`):
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: ClusterVersion
+metadata:
+  name: version
+spec:
+  desiredUpdate:
+    imageSource:
+      sourceType: Bundle
+```
+
+The technician that performs the upgrade will receive the USB stick containing
+the upgrade bundle and will plug it in one of the nodes of the cluster.
+
+The cluster version operator will be monitoring device events in all the nodes
+of the cluster in order to detect when such an USB stick (or any other kind of
+media containing an upgrade bundle) has been plugged. To control that a new
```

Comment:
CVO only runs in the control plane. So it wont be able to detect the USB if it is plugged in to the worker node.

Reply:
We are suggesting to run a `bundle-monitor` daemon set for this purpose.